tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -7630,16 +7630,78 @@ void SortOp::getCanonicalizationPatterns(RewritePatternSet& results,
 // TransposeOp
 //===----------------------------------------------------------------------===//
 
+template <typename T>
+static DenseElementsAttr foldTransposeHelper(mhlo::TransposeOp op,
+                                             DenseElementsAttr valueAttr) {
+  llvm::SmallVector<int64_t> permutation =
+      llvm::to_vector(op.getPermutation().getValues<int64_t>());
+  int64_t rank = permutation.size();
+  auto inputShape = op.getOperand().getType().cast<ShapedType>().getShape();
+  auto outputType = op.getType().cast<ShapedType>();
+  auto outputShape = outputType.getShape();
+
+  assert(outputType.hasStaticShape() && "output shape must be static");
+  assert(rank >= 2 && "permutation size must >= 2");
+  llvm::SmallVector<int64_t> strides(rank, 1);
+  llvm::SmallVector<int64_t> outputStrides(rank, 1);
+  llvm::SmallVector<int64_t> outputPermutation(rank, -1);
+  for (int64_t i = rank - 2; i >= 0; i--) {
+    strides[i] = strides[i + 1] * inputShape[i + 1];
+    outputStrides[i] = outputStrides[i + 1] * outputShape[i + 1];
+  }
+  for (int64_t i = 0; i < rank; i++) {
```

Comment:
These uses of `outputPermutation` and later related indexing logic need some more comments explaining what they are doing. This reversing of the permutation and, the `outputIndices` and `inputIndex` right now can only be understood by staring at the code for a while.